tensorflow_probability/python/distributions/johnson_su.py

```diff
@@ -0,0 +1,287 @@
+      dtype_util.assert_same_float_dtype((self._gamma, self._delta,
+                                          self._loc, self._scale))
+
+      norm_shift = invert_bijector.Invert(
+          shift_bijector.Shift(shift=self._gamma, validate_args=validate_args)
+      )
+
+      norm_scale = invert_bijector.Invert(
+          scale_bijector.Scale(scale=self._delta, validate_args=validate_args)
+      )
+
+      sinh = inline_bijector.Inline(
+          forward_fn=tf.sinh,
+          inverse_fn=tf.asinh,
+          forward_log_det_jacobian_fn=lambda x: tf.math.log(tf.math.cosh(x)),
+          inverse_log_det_jacobian_fn=lambda y: -0.5 * math.log1psquare(y),
+          forward_min_event_ndims=0,
+          is_increasing=lambda: True,
+          validate_args=validate_args,
+          name='sinh'
+      )
+
+      scale = scale_bijector.Scale(scale=self._scale,
+                                   validate_args=validate_args)
+
+      shift = shift_bijector.Shift(shift=self._loc,
+                                   validate_args=validate_args)
+
+      bijector = shift(scale(sinh(norm_scale(norm_shift))))
+
+      batch_shape = distribution_util.get_broadcast_shape(
+          self._gamma, self._delta, self._loc, self._scale)
+
+      super(JohnsonSU, self).__init__(
+          distribution=normal.Normal(
+              loc=tf.zeros([], dtype=dtype),
+              scale=tf.ones([], dtype=dtype),
+              validate_args=validate_args,
+              allow_nan_stats=allow_nan_stats),
+          bijector=bijector,
+          batch_shape=batch_shape,
+          validate_args=validate_args,
+          parameters=parameters,
+          name=name)
+
+  @staticmethod
+  def _param_shapes(sample_shape):
+    return dict(
+        zip(('gamma', 'delta', 'loc', 'scale'),
+            ([tf.convert_to_tensor(sample_shape, dtype=tf.int32)] * 4)))
+
+  @classmethod
+  def _params_event_ndims(cls):
+    return dict(gamma=0, delta=0, loc=0, scale=0)
+
+  @property
+  def gamma(self):
+    """Gamma shape parameters in these Johnson's SU distribution(s)."""
+    return self._gamma
+
+  @property
+  def delta(self):
+    """Delta shape parameters in these Johnson's SU distribution(s)."""
+    return self._delta
+
+  @property
+  def loc(self):
+    """Locations of these Johnson's SU distribution(s)."""
+    return self._loc
+
+  @property
+  def scale(self):
+    """Scaling factors of these Johnson's SU distribution(s)."""
+    return self._scale
+
+  def _mean(self):
+    gamma, delta, scale, loc = [tf.convert_to_tensor(v)
+                                for v in (self.gamma, self.delta, self.scale,
+                                          self.loc)]
+
+    return loc - scale * tf.math.exp(0.5 / tf.math.square(delta)) * \
+      tf.math.sinh(gamma / delta)
+
+  def _variance(self):
+    gamma, delta, scale = [tf.convert_to_tensor(v)
+                           for v in (self.gamma, self.delta, self.scale)]
+
+    inv_delta = 1./delta
```

Comment:
I would just inline these. We generally @tf.function wrap our code in the end, so that the computation will only be done once (because of common subexpression elimination in graph mode).

Reply:
Some test was failing because of this, now everything seems to be fine. Inlined.